packages/core/src/indexing/migrations/1-create-model-table.ts

```diff
@@ -214,6 +222,42 @@ export async function createSqliteModelTable(
   })
 }
 
+export async function createPostgresIndices(
+  dataSource: Knex,
+  tableName: string,
+  indices: Array<ModelIndex>
+): Promise<void> {
+  for (const index of indices) {
+    const json_indices = []
+    for (const field of index.fields) {
+      const path = field.path.map((p) => `'${p}'`)
+      json_indices.push(`(stream_content->${path.join('->')})`)
+    }
+    const indexName = modelIndexName(index, tableName)
+    await dataSource.schema.raw(
+      `CREATE INDEX ${indexName} ON ${tableName} (${json_indices.join(',')})`
+    )
+  }
+}
+
+export async function createSqliteIndices(
+  dataSource: Knex,
+  tableName: string,
+  indices: Array<ModelIndex>
+): Promise<void> {
+  for (const index of indices) {
+    const json_indices = []
+    for (const field of index.fields) {
+      const path = field.path.join('.')
+      json_indices.push(`json_extract(stream_content, '$.${path}')`)
+    }
+    const indexName = modelIndexName(index, tableName)
+    await dataSource.schema.raw(
```

Comment:
What is the expected behavior when running this query please?
What happens if the DBMS has millions of rows to index, does this call hang? Or on the contrary, does it resolve before the DB is fully indexed?

Reply:
As implemented currently, we only apply the index when a MID table is created, since models are immutable. The indices cannot be changed once the MID table is created. We will have to revisit this when models aren't immutable.

Reply:
We don't currently have anything that updates tables in js-ceramic. My inclination on this is to not add that functionality until needed, due to three reasons:
* We have immutable models, so developers would likely be creating new models or wrapping old ones, rather than creating an update on an existing model
* The indices do not prevent querying, only improve performance.
* ComposeDB can use the indices for querying/filtering even if they're not applied in js-ceramic

Reply:
That seems like a pretty big departure from what we described in the RFC, notably the ability to add query filters on existing models and the fact the ComposeDB client would generate the query filters types based on the indices set in the composites rather than any field.
To me that would make the whole feature much harder to explain and less useful to developers, cc @oed @stbrody any thoughts here please?

Reply:
> The indices do not prevent querying, only improve performance.

When we discussed this on a call the other day I thought we said that indices are actually required or else the GraphQL runtime will reject the query? 

> ComposeDB can use the indices for querying/filtering even if they're not applied in js-ceramic

I don't follow what this means?

> That seems like a pretty big departure from what we described in the RFC, notably the ability to add query filters on existing models

Yeah it was also my understanding that we would support adding new indices to existing Models, which is why I requested in my other comment a document describing how that would work.

Reply:
> When we discussed this on a call the other day I thought we said that indices are actually required or else the GraphQL runtime will reject the query?

That is correct. That will be implemented in js-composedb though.

>> ComposeDB can use the indices for querying/filtering even if they're not applied in js-ceramic
> I don't follow what this means?

ComposeDB can use the indices to create the necessary input types for querying/ordering. This PR isn't even necessary to do filtering/ordering, other than not adding the indices in js-ceramic would have a performance impact. The main point of this PR is to make sure that what is implemented in composedb works for communicating indices to js-ceramic and keep them outside of the model.

Reply:
I think the main problems here are the gap between expectations and implementation behavior, and overall usability of the feature.

To summarize the issue, here is the current behavior in ComposeDB v0.4:

- Deploying composites is idempotent, if the models are already indexed on a node, no error is thrown
- Composites are mutable, models can be added to them and views can be added to existing models (at the composite level)

Now from my understanding of the changes here, we’d have:

- Idempotent composites deployment: still possible or throwing an error if fields indices are set?
- No possibility to add indices to already deployed models
- No possibility to add views to already deployed models if they contain indices (due to the previous constraint)

Ideally the behavior we'd want is:

- Idempotent composites deployment: whether models already exist or not, and field indices need to be added or not, the composite deployment logic should work and not throw errors of existing indices
- Ability to add indices to existing composites (with already deployed models)

So then when the ComposeDB client generates the supported query filters types, there should be no "gap" in functionalities/expectations between what the client queries and the Ceramic node can efficiently resolve, as long as the composite has been deployed to the node.

Reply:
@PaulLeCam Please add this information to the project page so we can capture requirements for additional PR's to consider the feature completed.